### PR TITLE
replaced klog.Warningf with an error message in client_go

### LINKED ISSUE #94428: client-go prints a klog warning by default in library code which is not allowed
**What happened**:

A code change was merged that used `klog.Warningf` in library code by default (`NewDefaultClientLoadingRules` changed in a way that impacted all code in the world which uses that default, which is pretty common), which is not allowed by convention in Kube because client libraries may be used in UIs that wish to control output (klog is not consistent with other output).  The code change should have had a callback that allowed clients to optionally decide to print, not print to Warningf by default.

https://github.com/kubernetes/kubernetes/pull/78185#issuecomment-685765692

We should also document the convention for other reviewers that logging by default in client libraries is disallowed.

For instance, a callback would have allowed kubectl to print a human readable message `warning: KUBECONFIG does not exist` in general, and in specific cases (like `kubectl config` or login flows where KUBECONFIG not existing is totally reasonable) not print a message.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by Debanitrkl**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
`client-go` has several instances of `klog.Warningf.`
So, turned these warnings into actual errors that are returned by the functions.

#### Which issue(s) this PR fixes:

Fixes #94428 

```release-note
The klog.Warningf is replaced with an error message in client.go
```

#### Special notes for your reviewer:
Already sent a PR but got the branch deleted #104852

/cc @dims

### CODE REVIEW

**Comment by MadhavJivrajani** [comment]
@Debanitrkl thanks for updating the PR, I think we will need a release note here considering we are moving from `klog.Warningf` to either a callback or `klog.Info`.
/triage accepted
/assign

**Comment by MadhavJivrajani** [comment]
/hold
For squashing commits post reviews.

File: staging/src/k8s.io/client-go/tools/clientcmd/loader_test.go [hunk 1]
```diff
@@ -121,14 +123,56 @@ func TestNonExistentCommandLineFile(t *testing.T) {
...
+	flags.Set("logtostderr", "false")
+	flags.Set("v", "1")
+	klog.SetOutput(buffer)
+
 	_, err := loadingRules.Load()
 	if err != nil {
 		t.Fatalf("Unexpected error: %v", err)
 	}
+
+	klog.Flush()
```
Comment (MadhavJivrajani) [inline staging/src/k8s.io/client-go/tools/clientcmd/loader_test.go:170]:
```suggestion
	klog.Flush()
```
nit

File: staging/src/k8s.io/client-go/tools/clientcmd/loader_test.go [hunk 2]
```diff
@@ -121,14 +123,56 @@ func TestNonExistentCommandLineFile(t *testing.T) {
...
+		Precedence:       []string{"bogus1", "bogus2", "bogus3"},
+		WarnIfAllMissing: true,
+		Warner:           klog.Warning,
 	}
 
+	buffer := &bytes.Buffer{}
+
+	flags := &flag.FlagSet{}
+	klog.InitFlags(flags)
+	flags.Set("logtostderr", "false")
```
Comment (MadhavJivrajani) [inline staging/src/k8s.io/client-go/tools/clientcmd/loader_test.go:137]:
```suggestion
	klog.LogToStderr(false)
```

Comment (MadhavJivrajani) [reply staging/src/k8s.io/client-go/tools/clientcmd/loader_test.go]: Please make the change for the other test as well :)

File: staging/src/k8s.io/client-go/tools/clientcmd/loader_test.go [hunk 3]
```diff
@@ -121,14 +123,49 @@ func TestNonExistentCommandLineFile(t *testing.T) {
...
+
+func TestWarningMissingFiles(t *testing.T) {
+	envVarValue := "bogus"
+	os.Setenv(RecommendedConfigPathEnvVar, envVarValue)
+	loadingRules := NewDefaultClientConfigLoadingRules()
+
+	buffer := &bytes.Buffer{}
+
+	klog.LogToStderr(false)
+	klog.SetOutput(buffer)
```
Comment (MadhavJivrajani) [inline staging/src/k8s.io/client-go/tools/clientcmd/loader_test.go:157]:
```suggestion
	flags := &flag.FlagSet{}
	klog.InitFlags(flags)
	flags.Set("v", "1")
	klog.LogToStderr(false)
	klog.SetOutput(buffer)
```

**Review by MadhavJivrajani — CHANGES_REQUESTED** [review]
The unit test failure is because `flag` package is imported but not used. Once fixed, you will still get a failure because the `TestWarningMissingFiles` test requires a level to be set. PTAL at the suggestion :)
Please also try and run these two tests locally to make sure things work as expected.

**Comment by MadhavJivrajani** [comment]
@Debanitrkl now all that's left is squashing commits and adding a release note and we should be good to go!

**Comment by Debanitrkl** [comment]
tide/merge-method-squash

**Comment by MadhavJivrajani** [comment]
/assign @yliaog @caesarxuchao 
for approval

File: staging/src/k8s.io/client-go/tools/clientcmd/loader.go [hunk 4]
```diff
@@ -218,7 +232,8 @@ func (rules *ClientConfigLoadingRules) Load() (*clientcmdapi.Config, error) {
 	}
 
 	if rules.WarnIfAllMissing && len(missingList) > 0 && len(kubeconfigs) == 0 {
-		klog.Warningf("Config not found: %s", strings.Join(missingList, ", "))
+		missingString := strings.Join(missingList, ", ")
+		rules.Warner.Warn(fmt.Sprintf("Config not found: %s", missingString))
```
Comment (yliaog) [inline staging/src/k8s.io/client-go/tools/clientcmd/loader.go:236]: nit: missingString is only used here, its value could be embedded here directly.

File: staging/src/k8s.io/client-go/tools/clientcmd/loader_test.go [hunk 5]
```diff
@@ -121,14 +123,52 @@ func TestNonExistentCommandLineFile(t *testing.T) {
...
+	envVarValue := "bogus"
+	os.Setenv(RecommendedConfigPathEnvVar, envVarValue)
+	loadingRules := NewDefaultClientConfigLoadingRules()
+
+	buffer := &bytes.Buffer{}
+
+	flags := &flag.FlagSet{}
+	klog.InitFlags(flags)
+	flags.Set("v", "1")
+	klog.LogToStderr(false)
```
Comment (yliaog) [inline staging/src/k8s.io/client-go/tools/clientcmd/loader_test.go:159]: this test confirms that the message is printed out when v=1 is set, but it does not confirm the message is *not* printed out when v=1 is not set.

**Pull request closed without merging**